### PR TITLE
C#: offload work from GrpcThreadPool by default.

### PR DESCRIPTION
**Pull request by jtattermusch**
Tentative fix for https://github.com/grpc/grpc/issues/11824

There's noticeable performance impact (600 -> 380KQPS  ping pong roundtrips on 32core), but offloading to a different thread has some extra safety features (it's hard to deadlock youself) and might even increase throughput in scenarios where it takes long time to serialize/deserialize requests or processing them is CPU intensive. Adding an experimental knob to allow turning off the offloading for those who really want to.

### CODE REVIEW

**Comment by jtattermusch** [comment]
CC @jskeet and @benwulfe  as this should fix #11824.

CC @chwarr I remember you mentioning the issue this is trying to address.

**Comment by benwulfe** [comment]
If a request still needs to hit the grpc thread before being offloaded,
then it wouldn't fix it.

**Comment by benwulfe** [comment]
Oh this should hide grpc thread from the outside world another way. Lgtm
then and thanks for fixing!

File: src/csharp/Grpc.Core/Internal/GrpcThreadPool.cs [hunk 1]
```diff
@@ -165,11 +170,19 @@ private void RunHandlerLoop(CompletionQueueSafeHandle cq, IProfiler optionalProf
                     try
                     {
                         var callback = cq.CompletionRegistry.Extract(tag);
-                        callback(success);
+                        // Use cached delegates to avoid unnecessary allocations
+                        if (!inlineHandlers)
+                        {
+                            ThreadPool.QueueUserWorkItem(success ? RunCompletionQueueEventCallbackSuccess : RunCompletionQueueEventCallbackFailure, callback);
```
Comment (jskeet) [inline src/csharp/Grpc.Core/Internal/GrpcThreadPool.cs]: Have you measured the performance impact of this vs `Task.Run`? I'm not sure whether `ThreadPool.QueueUserWorkItem` uses the same task-stealing queues. But we could change it later in a non-breaking way, of course.

**Comment by benwulfe** [comment]
Hey Jan, I know you had previously suggested using fewer completion queues to fix this.  Is there any advantage to having more than **one** completion queue?  It seems that with a single queue serviced by all threads, you achieve more fairness (and far lower deadlock chances).  With multiple queues, you could get an operation stuck indefinitely behind a fairly long operation while the other queues are happily processed, right?

This change lgtm -- but if this change has drawbacks, maybe the immediate fix would be to make the default value of  completionqueuecount = "1"?

**Comment by apolcyn** [comment]
One comment:
It seems like issues with the current thread pool are not just for <i>sync unary RPC types</i>. But if getting rid of the deadlock issue for sync RPC types was enough, then could we use `grpc_completion_queue_pluck` for these call types and completely avoid the thread pool for them, in order to keep the current thread pool for it's performance?

File: src/csharp/Grpc.Core/Internal/GrpcThreadPool.cs (hunk 1 above)
Comment (jtattermusch) [inline src/csharp/Grpc.Core/Internal/GrpcThreadPool.cs]: I just tried with Task.Run() and the performance seems about the same, but the advantage of ThreadPool.QueueUserWorkItem is that it saves allocations, which might become interesting in the future when we look into optimizing the performance a bit more (I already know about some other places where we allocate unnecessarily and I've tested that removing those allocations will improve the performance).

File: src/csharp/Grpc.Core/GrpcEnvironment.cs [hunk 2]
```diff
@@ -218,12 +219,31 @@ public static void SetCompletionQueueCount(int completionQueueCount)
         }
 
         /// <summary>
+        /// By default, gRPC's internal event handlers get offloaded to .NET default thread pool thread (<c>inlineHandlers=false</c>).
+        /// Setting <c>inlineHandlers</c> to <c>true</c> will allow scheduling the event handlers directly to
```
Comment (chwarr) [inline src/csharp/Grpc.Core/GrpcEnvironment.cs]: Minor: the use of `<c>` around `inlineHandlers` seems intentional here, but so you are aware, there's also [`<paramref>`](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/xmldoc/paramref).

**Pull request merged**